Rewrite autoloader so we have all rules defined in separate class

michalbundyra opened the pull request:
The class with defined rules now is reusable and we can use it also on rewriting the libraries

~We need to think what we can do with functions, and somehow I think it will be better to create these alias in repositories where we have defined functions.~

~We would need define list of all possible function defined in libraries and then just call new function using `call_user_func_array` and `func_get_args`.~

**UPDATE:** As discussed, we will try to handle functions separately on rewriting packages, as these are only in the following pacakges:
- zend-diactoros
- zend-stragitility
- zend-container-config-test

michalbundyra committed:
Rewrite autoloader so we have all rules defined in separate class

The class with defined rules now is reusable and we can use it
also on rewriting the libraries

We need to think what we can do with functions,
and somehow I think it will be better to create these alias
in repositories where we have defined functions.

We would need define list of all possible function defined in
libraries and then just call new function using `call_user_func_array`
and `func_get_args`.

michalbundyra committed:
Removes function aliasing

We should handle it separetely on rewriting packages

michalbundyra committed:
Better detection of namespace to rewrite

michalbundyra committed:
Changed to long array notation as we want to support also old PHP versions

michalbundyra committed:
Align key => value pairs in array

weierophinney commented on src/Autoloader.php:
```diff
@@ -0,0 +1,39 @@
+                ++$i;
+            }
+
+            if ($check !== '') {
```
I'd invert this condition:

```php
if ($check === '') {
    return;
}
```

weierophinney commented on src/Autoloader.php:
```diff
@@ -0,0 +1,39 @@
+            while (isset($classes[$check . $segments[$i] . '\\'])) {
+                $check .= $segments[$i] . '\\';
+                ++$i;
+            }
```
This seems really convoluted, honestly. Why not just:

```php
foreach ($class as $legacyPrefix => $aliasPrefix) {
    if (strpos($class, $legacyPrefix) === 0) {
        $alias = str_replace($legacyPrefix, $aliasPrefix, $class);
        class_alias($alias, $class);
        return;
    }
}
```

michalbundyra replied:
It was exactly like that before. Now we are using segments, so we can return quicker - in 1 or two steps. With foreach we need to check it in the worst case scenario 12 times...

weierophinney replied:
Yep - I was misreading it previously.

Maybe put in a comment about how it works, so others who come to it later know why it's written the way it is.

michalbundyra replied:
Comment is added now.

michalbundyra commented on src/RewriteRules.php:
```diff
@@ -0,0 +1,37 @@
+<?php
+/**
+ * @see       https://github.com/laminas/laminas-zendframework-bridge for the canonical source repository
+ * @copyright Copyright (c) 2019 Laminas Foundation (https://getlaminas.org)
```
we should update it now to refer to COPYRIGHT.md file now

weierophinney replied:
Yes, please. :)

michalbundyra committed:
Updated copyright tag in files to refer to COPYRIGHT.md

michalbundyra committed:
Swapped condition to return earlier

michalbundyra committed:
Added comment to clarify how matching namespaces works

We are using namespace segments instead of looping of the whole array
of the namespaces to rewrite. Because of that we are able to quit after
1 or 2 checks, otherwise - if we use normal loop - in the worst case
scenario - we need to have 12 checks.
